src/OpenLoco/src/GameCommands/CreateIndustry.cpp

```diff
@@ -0,0 +1,767 @@
+                elIndustry->setSectionProgress(0);
+                elIndustry->setColour(colour);
+                elIndustry->setBuildingType(buildingType);
+                elIndustry->setVar_6_003F(0);
+                World::AnimationManager::createAnimation(3, World::toWorldSpace(tilePos), elIndustry->baseZ());
+                elIndustry->setGhost(flags & Flags::flag_6);
+                Ui::ViewportManager::invalidate(World::toWorldSpace(tilePos), elIndustry->baseHeight(), elIndustry->clearHeight());
+            }
+        }
+
+        auto& tileRef = industry->tiles[industry->numTiles];
+        tileRef = World::Pos3(pos, (highestBaseZ * World::kSmallZStep | (isMultiTile ? 0x8000 : 0)));
+        industry->numTiles++;
+        return totalCost;
+    }
+
+    // 0x0045436B
+    static currency32_t createIndustry(const IndustryPlacementArgs& args, const uint8_t flags)
+    {
+        GameCommands::setExpenditureType(ExpenditureType::Miscellaneous);
+        {
+            const auto centrePos = World::Pos2(args.pos.x + 16, args.pos.y + 16);
+            const auto centreHeight = World::TileManager::getHeight(centrePos);
+            GameCommands::setPosition({ args.pos.x, args.pos.y, centreHeight.landHeight });
+        }
+
+        // Vanilla bug args.srand0 not set. Change this when we can diverge
+        Core::Prng prng{ enumValue(GameCommand::createIndustry), args.srand1 };
+        const auto newIndustryId = sub_454C91(args.type, args.pos, prng);
+        if (newIndustryId == IndustryId::null)
+        {
+            return FAILURE;
+        }
+        _industryLastPlacedId = newIndustryId;
+        auto* newIndustry = IndustryManager::get(newIndustryId);
+        auto* indObj = newIndustry->getObject();
+        if (args.buildImmediately)
+        {
+            newIndustry->under_construction = 0xFF;
+        }
+        if (flags & Flags::flag_6)
+        {
+            newIndustry->flags |= IndustryFlags::isGhost;
+        }
+        if (!isEditorMode() && getUpdatingCompanyId() != CompanyId::neutral)
+        {
+            newIndustry->flags |= IndustryFlags::flag_04;
+            newIndustry->owner = getUpdatingCompanyId();
+        }
+
+        // Note: Could use a fixed size vector as max size is Colour::max
+        std::vector<Colour> availableColours;
+        auto colourBitSet = indObj->availableColours;
+        for (auto colourI32 = Numerics::bitScanForward(colourBitSet); colourI32 != -1; colourI32 = Numerics::bitScanForward(colourBitSet))
+        {
+            colourBitSet &= ~(1ULL << colourI32);
+            availableColours.push_back(static_cast<Colour>(colourI32));
+        }
+
+        // 0x00E0C3D3
+        const auto randColour = [&availableColours, &newIndustry]() {
+            if (availableColours.empty())
+            {
+                return Colour::black;
+            }
+            return availableColours[newIndustry->prng.randNext(availableColours.size() - 1)];
+        }();
```

Comment:
I would probably move this to a new function, the first part just collects the available colors, the second part selects one, so something like `static Colour getRandomColourForIndustry(const Industry& industry)`